### DISCUSSION TITLE
📎 Split up xtask into smaller crates

### DISCUSSION THREAD

**Discussion by Boshen** [rome/tools#1947] [Suggestions]
### Description

The `xtask` crate is getting bloated. This has the side effect of:
* super slow compilation speed when running xtask commands, since compilation is cached by unit of crates, splitting it up will improve both local and CI build time. 
* our codegen scripts fail if any crate that the parser depends on has a syntax error. This is particularly annoying if there are merge conflicts in generated files

---

Rust Compilation speed needs to be engineered, development experience can suffer if we are not careful. We can use `cargo bloat` to watch out for outliers.

On my 2015 mac book, Running `cargo bloat --time -j 1` gives
```
102.37s lsp_types
 75.87s rslint_parser
 73.65s clap
 72.18s rome_lsp
 42.39s criterion
 40.82s lspower
 36.47s futures_util
 31.71s tokio
 29.53s clap_derive
 29.31s rayon
 28.81s xtask
 27.40s serde_derive
 26.97s schemars
 25.64s syn
 23.27s syn
 20.24s rustls
 20.03s clap
 ... more omitted
```

xtask and its dependencies are all sitting on the top.

**Comment by ematipico** [comment]
How would you suggest to split them up?

**Reply by Boshen** [reply]
Along the lines of 
```
xtask/
  cli/
  codegen/
  syntax/
  format/
  compare/
  coverage
```
\+ Some command aliasing in .cargo/config.toml

**Reply by ematipico** [reply]
If we decided to move each command in its own crate, I don't think it would make sense to have a CLI. I mean, the CLI is what glues all together, resulting on a bloat.

I think we can just create cargo aliases and call it a day? And we could prefix those aliases with `cargo dev`?

**Reply by Boshen** [reply]
I'll experiment on this once all the other PRs are merged in (rust 1.58.0, whitespace, unicode codegen).

**Comment by Boshen** [comment]
I'm going to start working on this. I'm currently waiting forever for everything to compile in xtask 😢

**Reply by ematipico** [reply]
Would you mind opening issue first? And explain how you plan to implement the solution? I believe there's room for multiple "good first issue" once we know how to move each command.

**Reply by Boshen** [reply]
I'm almost done moving them 😄 It's easier to untangle everything in one piece. I'll create a Draft PR soon.
